minio/thread_pool.py

```diff
@@ -42,20 +42,25 @@ def __init__(self, tasks_queue, results_queue, exceptions_queue):
         self.start()
 
     def run(self):
-        fast_quit = False
-        while not self.tasks_queue.empty():
-            func, args, kargs = self.tasks_queue.get()
-            if not fast_quit:
+        """ Continously receive tasks and execute them """
+        while True:
+            task = self.tasks_queue.get()
+            if task is None:
+                self.tasks_queue.task_done()
+                break
+            # No exception detected in any thread,
+            # continue the execution.
+            if self.exceptions_queue.empty():
```

Comment:
I suspect a small contingency here, I doubt that the single worker is fed with all the tasks in the task queue here.
This condition `self.exceptions_queue.empty()` checks for the exception queue alone and allows all the tasks to actually reach `result = func(*args, **kargs)`. I can break out in points here
- Initially we have 3 worker threads up and ready by `pool.start_parallel()`
- Then we add the tasks one by one by `pool.add_task()`
- This condition here `self.exceptions_queue.empty()` allows any task to succeed and reach `result = func(*args, **kargs)`. Say, If there are 3 tasks and this condition here will pass until any exception is thrown yet. So any single worker will accept all the tasks. 
I think there should be something like `while not self.tasks_queue.empty()` so that the decision is made up on task_queue instead of exception_queue
 
I can be wrong or out of clarity. So please correct me if i'm wrong.

Reply:
> * Say, If there are 3 tasks and this condition here will pass until any exception is thrown yet. So any single worker in the pool will accept all the tasks.

I didn't fully understand your point, let's talk privately

Reply:
Clear now .. Thanks Anis!